src/mlpack/methods/ann/rbm_impl.hpp

```diff
@@ -0,0 +1,266 @@
+{
+  arma::mat output;
+  visible.ForwardPreActivation(std::move(input), std::move(output));
+  SoftplusFunction::Fn(output, output);
+  return  -(arma::accu(output) + arma::dot(input, visible.Bias()));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Evaluate(const arma::mat& /* parameters*/, const size_t i)
+{
+  arma::mat tempInput;
+  arma::vec currentInput = predictors.unsafe_col(i);
+  if (!useMonitoringCost)
+  {
+    // Do not use unsafe col predictor as we change the input
+    Gibbs(i, std::move(tempInput));
+    return FreeEnergy(currentInput) - FreeEnergy(tempInput);
+  }
+  else
+  {
+    if(persistence)
+    {
+      size_t idx = RandInt(0, currentInput.n_rows);
+      tempInput = currentInput;
+      tempInput.row(idx) = 1 - currentInput.row(idx);
+      // Use mean when showing the if more than one input
+      return std::log(LogisticFunction::Fn(FreeEnergy(tempInput) -
+          FreeEnergy(currentInput))) * predictors.n_rows;
+    }
+    else
+    {
+      visible.Forward(std::move(predictors.col(i)), std::move(tempInput));
+      // Use mean when showing the if more than one input
+      return arma::accu( predictors.col(i) * arma::log(tempInput).t() +
+          (1 - predictors.col(i)) * arma::log(1 - tempInput).t());
+    }
+  }
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    SampleHidden(arma::mat&& input, arma::mat&& output)
+{
+  visible.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    SampleVisible(arma::mat&& input, arma::mat&& output)
+{
+  hidden.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Gibbs(const size_t i, arma::mat&& output, size_t steps)
+{
+  // Do not use unsafe col
+  gibbsInput = predictors.col(i);
+  steps = (steps == 1)?this-> numSteps: steps;
+  if (persistence && !state.is_empty())
+    gibbsInput = state;
+
+  if (steps == 1)
+  {
+    this->VisibleLayer().Sample(std::move(gibbsInput), std::move(output));
+    this->HiddenLayer().Sample(std::move(output), std::move(output));
+  }
+  else
+  {
+    for (size_t j = 0; j < steps - 1 ; j++)
+    {
+      // Use probabilties for updation till the last step(section 3 hinton)
+      SampleHidden(std::move(gibbsInput), std::move(output));
+      SampleVisible(std::move(output), std::move(gibbsInput));
+    }
+    // Set state to binarised output
+    this->VisibleLayer().Forward(std::move(gibbsInput), std::move(state));
```

Comment:
Okay if you look closely at the deep learning.net example the gibbshvh to do the negative pass where you do this SampleVisible ---> SampleHidden(Negative Hidden) ---> SampleVisible(Negative Visible)x'
Also state = NegativeHidden and output = Negative Visible.
The same strategy cannot be applied for Sampling for sampling they use 
gibbsvhv where they do something like 
SampleHidden--> VisibleForwardPass --> SampleVisible(
here output = Visible.ForwardPass(input) and state = SampleVisible.
That was the mistake we were making.